### PR TITLE
Some support for cis/trans arithmetic

### PR DESCRIPTION
**Pull request by Phlya**
Coverage saves total cis contacts into info (when `store=True`).
Sample can have total cis contacts as target: tries to use the stored value and calculates it if it's not there.

### CODE REVIEW

File: cooltools/api/coverage.py [hunk 1]
```diff
@@ -119,5 +119,5 @@ def coverage(
                     del grp["bins"][store_name]
                 h5opts = dict(compression="gzip", compression_opts=6)
                 grp["bins"].create_dataset(store_name, data=cov_arr, **h5opts)
-
+            grp.attrs.create("cis", np.sum(covs[0]), dtype=int)
```
Comment (sergpolly) [inline cooltools/api/coverage.py:122]: just make sure it goes to cooler.info
otherwise it is super cool

File: cooltools/api/sample.py [hunk 2]
```diff
@@ -71,6 +73,10 @@ def sample(
         The target sample size as a fraction of contacts in the original
         dataset. Mutually exclusive with `count`.
 
+    cis_target : bool
```
Comment (sergpolly) [inline cooltools/api/sample.py:76]: maybe rename to cis_count and move up to be close to the original `count`

File: cooltools/api/sample.py [hunk 3]
```diff
@@ -88,7 +94,11 @@ def sample(
         clr = cooler.Cooler(clr)
 
     if count is not None and frac is None:
```
Comment (sergpolly) [inline cooltools/api/sample.py:96]:
```python
# mutually exclusive frac, count and cis_count
if frac is not None and count is None and cis_count is None:
    # work with frac
elif frac is None and count is not None and cis_count is None:
   # work with count
elif frac is None and count is None and cis_count is not None:
   # work with cis_count
else:
    raise
```

File: cooltools/api/sample.py [hunk 4]
```diff
@@ -48,6 +49,7 @@ def sample(
     out_clr_path,
     count=None,
     frac=None,
+    cis_target=False,
     exact=False,
```
Comment (sergpolly) [inline cooltools/api/sample.py:53]: consider adding `ignore_diags` here ? or ultimately `dist_min` and `dist_max` ?
if it does not end up oin this PR - leave an issue,  maybe ? for the future

File: cooltools/api/coverage.py (hunk 1 above)
Comment (Phlya) [reply cooltools/api/coverage.py:122]: It does!

File: cooltools/api/sample.py (hunk 2 above)
Comment (Phlya) [reply cooltools/api/sample.py:80]: Done!

File: cooltools/api/sample.py (hunk 3 above)
Comment (Phlya) [reply cooltools/api/sample.py]: Thank you, done!

**Comment by Phlya** [comment]
All done except ignore_diags, also added tests for coverage API, and fixed a bug in exact sampling (which was broken for me due to wrong dtype).

**Comment by Phlya** [comment]
Apart from the idea about ignore_diags this is ready IMO

File: cooltools/api/sample.py (hunk 4 above)
Comment (sergpolly) [reply cooltools/api/sample.py:53]: I'd say yes ... Also just make sure what `ignore_diags` means exactly - ignore diags to set the count target, of course - but what should happen with the ignored_diags in the subsamples matrix ? Should they be zero ?

Comment (Phlya) [reply cooltools/api/sample.py:53]: Yes meaning ... which of the options?
I'd say the calculation of the subsampling fraction should not take `ignore_diags` diags into account, but subsampling would be applied to all data including those diags

File: cooltools/api/sample.py [hunk 5]
```diff
@@ -63,13 +65,21 @@ def sample(
...
+    cis_count : int
+        The target number of cis contacts in the sample.
+        Mutually exclusive with `count` and `frac`.
 
     frac : float
         The target sample size as a fraction of contacts in the original
-        dataset. Mutually exclusive with `count`.
+        dataset. Mutually exclusive with `count` and `cis_count`.
+
+    cis_target : bool
```
Comment (sergpolly) [inline cooltools/api/sample.py:80]: remove `cis_target`  from docstring ?

File: tests/test_sample.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        op.join(request.fspath.dirname, "data/CN.mm9.1000kb.test_sampled.cool")
+    )
+    # Test that deviation from expected total is very small
+    testing.assert_allclose(clr_result.info["sum"], 200000000, rtol=1e-3)
+
+
+# Exact sampling is very slow! So commented out
+# cooltools.api.sample.sample(
+#     clr,
+#     op.join(request.fspath.dirname, "data/CN.mm9.1000kb.test_sampled.cool"),
```
Comment (sergpolly) [inline tests/test_sample.py:39]: can we use a smaller input sample to uncomment this ?

Comment (Phlya) [reply tests/test_sample.py:39]: This is our normal test cooler at 1Mb resolution... So quite small! Idk we could add a 10Mb resolution one?

Comment (Phlya) [reply tests/test_sample.py:39]: Even with 10Mb resolution it takes a while... This whole file takes 95 sec to run through pytest on my laptop, which I guess is acceptable?

Comment (sergpolly) [reply tests/test_sample.py]: let's try this way - w'd need to optimize out tests at some point anyways

**Pull request merged**